File: app/src/main/java/com/elementary/tasks/birthdays/ShowBirthdayActivity.java
package com.elementary.tasks.birthdays;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import com.elementary.tasks.R;
import com.elementary.tasks.core.BaseNotificationActivity;
import com.elementary.tasks.core.async.BackupTask;
import com.elementary.tasks.core.utils.Configs;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Contacts;
import com.elementary.tasks.core.utils.LED;
import com.elementary.tasks.core.utils.Language;
import com.elementary.tasks.core.utils.Module;
import com.elementary.tasks.core.utils.Notifier;
import com.elementary.tasks.core.utils.Permissions;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.Sound;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.utils.TelephonyUtil;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.core.utils.ViewUtils;
import com.elementary.tasks.core.views.roboto.RoboTextView;
import com.elementary.tasks.databinding.ActivityShowBirthdayBinding;

import java.util.Calendar;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.databinding.DataBindingUtil;
import de.hdodenhof.circleimageview.CircleImageView;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class ShowBirthdayActivity extends BaseNotificationActivity {

    private static final int CALL_PERM = 612;
    private static final int SMS_PERM = 613;

    private ActivityShowBirthdayBinding binding;
    @Nullable
    private BirthdayItem mBirthdayItem;
    private boolean mIsResumed;
    @Nullable
    private String wearMessage;

    public static Intent getLaunchIntent(Context context, String uuId) {
        Intent resultIntent = new Intent(context, ShowBirthdayActivity.class);
        resultIntent.putExtra(Constants.INTENT_ID, uuId);
        resultIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
        return resultIntent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        mIsResumed = getIntent().getBooleanExtra(Constants.INTENT_NOTIFICATION, false);
        mBirthdayItem = RealmDb.getInstance().getBirthday(getIntent().getStringExtra(Constants.INTENT_ID));
        super.onCreate(savedInstanceState);
        if (mBirthdayItem == null) finish();
        binding = DataBindingUtil.setContentView(this, R.layout.activity_show_birthday);
        binding.card.setCardBackgroundColor(getThemeUtil().getCardStyle());
        if (Module.isLollipop()) {
            binding.card.setCardElevation(Configs.CARD_ELEVATION);
        }
        loadImage(binding.bgImage);
        colorify(binding.buttonOk, binding.buttonCall, binding.buttonSend);

        binding.buttonOk.setOnClickListener(view -> ok());
        binding.buttonCall.setOnClickListener(view -> call());
        binding.buttonSend.setOnClickListener(view -> sendSMS());

        binding.buttonOk.setImageResource(R.drawable.ic_done_black_24dp);
        binding.buttonCall.setImageResource(R.drawable.ic_call_black_24dp);
        binding.buttonSend.setImageResource(R.drawable.ic_send_black_24dp);

        CircleImageView contactPhoto = binding.contactPhoto;
        contactPhoto.setBorderColor(getThemeUtil().getColor(getThemeUtil().colorPrimary()));
        contactPhoto.setVisibility(View.GONE);

        if (!TextUtils.isEmpty(mBirthdayItem.getNumber()) && checkContactPermission()) {
            mBirthdayItem.setNumber(Contacts.getNumber(mBirthdayItem.getName(), this));
        }
        if (mBirthdayItem.getContactId() == 0 && !TextUtils.isEmpty(mBirthdayItem.getNumber()) && checkContactPermission()) {
            mBirthdayItem.setContactId(Contacts.getIdFromNumber(mBirthdayItem.getNumber(), this));
        }
        Uri photo = Contacts.getPhoto(mBirthdayItem.getContactId());
        if (photo != null) {
            contactPhoto.setImageURI(photo);
        } else {
            contactPhoto.setVisibility(View.GONE);
        }
        String years = TimeUtil.getAgeFormatted(this, mBirthdayItem.getDate());
        RoboTextView userName = binding.userName;
        userName.setText(mBirthdayItem.getName());
        userName.setContentDescription(mBirthdayItem.getName());
        RoboTextView userNumber = findViewById(R.id.userNumber);
        RoboTextView userYears = findViewById(R.id.userYears);
        userYears.setText(years);
        userYears.setContentDescription(years);
        wearMessage = mBirthdayItem.getName() + "\n" + years;
        if (TextUtils.isEmpty(mBirthdayItem.getNumber())) {
            binding.buttonCall.hide();
            binding.buttonSend.hide();
            userNumber.setVisibility(View.GONE);
        } else {
            userNumber.setText(mBirthdayItem.getNumber());
            userNumber.setContentDescription(mBirthdayItem.getNumber());
        }
        showNotification(TimeUtil.getAge(mBirthdayItem.getDate()), mBirthdayItem.getName());
        if (isTtsEnabled()) {
            startTts();
        }
    }

    private boolean checkContactPermission() {
        return Permissions.checkPermission(this, Permissions.READ_CONTACTS, Permissions.READ_CALLS);
    }

    public void showNotification(int years, String name) {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, Notifier.CHANNEL_REMINDER);
        builder.setContentTitle(name);
        builder.setContentText(TimeUtil.getAgeFormatted(this, years));
        if (Module.isLollipop()) {
            builder.setSmallIcon(R.drawable.ic_cake_white_24dp);
            builder.setColor(ViewUtils.getColor(this, R.color.bluePrimary));
        } else {
            builder.setSmallIcon(R.drawable.ic_cake_nv_white);
        }
        if (!isScreenResumed() && (!SuperUtil.isDoNotDisturbEnabled(this) ||
                (SuperUtil.checkNotificationPermission(this) && isBirthdaySilentEnabled()))) {
            Sound sound = getSound();
            if (sound != null) getSound().playAlarm(getSoundUri(), isBirthdayInfiniteSound());
        }
        if (isVibrate()) {
            long[] pattern = new long[]{150, 400, 100, 450, 200, 500, 300, 500};
            if (isBirthdayInfiniteVibration()) {
                pattern = new long[]{150, 86400000};
            }
            builder.setVibrate(pattern);
        }
        if (Module.isPro()) {
            builder.setLights(getLedColor(), 500, 1000);
        }
        boolean isWear = getPrefs().isWearEnabled();
        if (isWear && Module.isJellyMR2()) {
            builder.setOnlyAlertOnce(true);
            builder.setGroup("GROUP");
            builder.setGroupSummary(true);
        }
        NotificationManagerCompat mNotifyMgr = NotificationManagerCompat.from(this);
        mNotifyMgr.notify(getId(), builder.build());
        if (isWear) {
            showWearNotification(name);
        }
    }

    private boolean isBirthdaySilentEnabled() {
        boolean is = getPrefs().isSoundInSilentModeEnabled();
        if (Module.isPro() && !isGlobal()) {
            is = getPrefs().isBirthdaySilentEnabled();
        }
        return is;
    }

    private boolean isTtsEnabled() {
        boolean is = getPrefs().isTtsEnabled();
        if (Module.isPro() && !isGlobal()) {
            is = getPrefs().isBirthdayTtsEnabled();
        }
        return is;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        removeFlags();
        if (getPrefs().isAutoBackupEnabled()) {
            new BackupTask(this).execute();
        }
        new BackupBirthdaysTask(this).execute();
    }

    @Override
    public void onBackPressed() {
        discardMedia();
        if (getPrefs().isFoldingEnabled()) {
            removeFlags();
            finish();
        } else {
            Toast.makeText(ShowBirthdayActivity.this, getString(R.string.select_one_of_item), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void call() {
        makeCall();
    }

    private void makeCall() {
        if (Permissions.checkPermission(this, Permissions.CALL_PHONE) && mBirthdayItem != null) {
            TelephonyUtil.makeCall(mBirthdayItem.getNumber(), this);
            updateBirthday(mBirthdayItem);
        } else {
            Permissions.requestPermission(this, CALL_PERM, Permissions.CALL_PHONE);
        }
    }

    @Override
    protected void delay() {
        close();
    }

    @Override
    protected void cancel() {
        sendSMS();
    }

    private void sendSMS() {
        if (Permissions.checkPermission(ShowBirthdayActivity.this, Permissions.SEND_SMS) && mBirthdayItem != null) {
            TelephonyUtil.sendSms(mBirthdayItem.getNumber(), ShowBirthdayActivity.this);
            updateBirthday(mBirthdayItem);
        } else {
            Permissions.requestPermission(ShowBirthdayActivity.this, SMS_PERM, Permissions.SEND_SMS);
        }
    }

    @Override
    protected void favourite() {
        close();
    }

    @Override
    protected void ok() {
        updateBirthday(mBirthdayItem);
    }

    private void updateBirthday(@Nullable BirthdayItem birthdayItem) {
        if (birthdayItem != null) {
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(System.currentTimeMillis());
            int year = calendar.get(Calendar.YEAR);
            birthdayItem.setShowedYear(year);
            RealmDb.getInstance().saveObject(birthdayItem);
        }
        close();
    }

    private void close() {
        removeFlags();
        discardNotification(getId());
        finish();
    }

    @Override
    protected void showSendingError() {
        binding.buttonCall.setImageResource(R.drawable.ic_refresh);
        binding.buttonCall.setContentDescription(getString(R.string.acc_button_retry_to_send_message));
        if (binding.buttonCall.getVisibility() == View.GONE) {
            binding.buttonCall.show();
        }
    }

    @Override
    protected Locale getTtsLocale() {
        Locale locale = new Language().getLocale(this, false);
        if (Module.isPro() && !isGlobal()) {
            locale = new Language().getLocale(this, true);
        }
        return locale;
    }

    @Override
    protected String getMelody() {
        if (Module.isPro() && !isGlobal()) {
            return getPrefs().getBirthdayMelody();
        } else {
            return getPrefs().getMelodyFile();
        }
    }

    @Override
    protected boolean isBirthdayInfiniteVibration() {
        boolean vibrate = getPrefs().isInfiniteVibrateEnabled();
        if (Module.isPro() && !isGlobal()) {
            vibrate = getPrefs().isBirthdayInfiniteVibrationEnabled();
        }
        return vibrate;
    }

    @Override
    protected boolean isBirthdayInfiniteSound() {
        boolean isLooping = getPrefs().isInfiniteSoundEnabled();
        if (Module.isPro() && !isGlobal()) {
            isLooping = getPrefs().isBirthdayInfiniteSoundEnabled();
        }
        return isLooping;
    }

    @Override
    protected boolean isScreenResumed() {
        return mIsResumed;
    }

    @Override
    protected boolean isVibrate() {
        boolean vibrate = getPrefs().isVibrateEnabled();
        if (Module.isPro() && !isGlobal()) {
            vibrate = getPrefs().isBirthdayVibrationEnabled();
        }
        return vibrate;
    }

    @Override
    protected String getSummary() {
        return wearMessage;
    }

    @Override
    protected String getUuId() {
        if (mBirthdayItem != null) {
            return mBirthdayItem.getUuId();
        } else return "";
    }

    @Override
    protected int getId() {
        if (mBirthdayItem != null) {
            return mBirthdayItem.getUniqueId();
        } else return 0;
    }

    @Override
    protected int getLedColor() {
        int ledColor = LED.getLED(getPrefs().getLedColor());
        if (Module.isPro() && !isGlobal()) {
            ledColor = LED.getLED(getPrefs().getBirthdayLedColor());
        }
        return ledColor;
    }

    @Override
    protected boolean isAwakeDevice() {
        boolean isWake = getPrefs().isDeviceAwakeEnabled();
        if (Module.isPro() && !isGlobal()) {
            isWake = getPrefs().isBirthdayWakeEnabled();
        }
        return isWake;
    }

    @Override
    protected int getMaxVolume() {
        return getPrefs().getLoudness();
    }

    @Override
    protected boolean isGlobal() {
        return getPrefs().isBirthdayGlobalEnabled();
    }

    @Override
    protected boolean isUnlockDevice() {
        return getPrefs().isDeviceUnlockEnabled();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length == 0) return;
        switch (requestCode) {
            case CALL_PERM:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    makeCall();
                }
                break;
            case SMS_PERM:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    sendSMS();
                }
                break;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/birthdays/ShowBirthdayActivity.kt
package com.elementary.tasks.birthdays

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.elementary.tasks.R
import com.elementary.tasks.birthdays.work.BackupBirthdaysTask
import com.elementary.tasks.core.BaseNotificationActivity
import com.elementary.tasks.core.async.BackupTask
import com.elementary.tasks.core.data.models.Birthday
import com.elementary.tasks.core.utils.*
import com.elementary.tasks.core.viewModels.Commands
import com.elementary.tasks.core.viewModels.birthdays.BirthdayViewModel
import com.mcxiaoke.koi.ext.onClick
import kotlinx.android.synthetic.main.activity_show_birthday.*
import java.util.*

/**
 * Copyright 2016 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class ShowBirthdayActivity : BaseNotificationActivity() {

    private var viewModel: BirthdayViewModel? = null
    private var mBirthday: Birthday? = null
    override var isScreenResumed: Boolean = false
        private set
    override var summary: String = ""
        private set

    private val isBirthdaySilentEnabled: Boolean
        get() {
            var `is` = prefs.isSoundInSilentModeEnabled
            if (Module.isPro && !isGlobal) {
                `is` = prefs.isBirthdaySilentEnabled
            }
            return `is`
        }

    private val isTtsEnabled: Boolean
        get() {
            var `is` = prefs.isTtsEnabled
            if (Module.isPro && !isGlobal) {
                `is` = prefs.isBirthdayTtsEnabled
            }
            return `is`
        }

    override val ttsLocale: Locale?
        get() {
            var locale = language.getLocale(false)
            if (Module.isPro && !isGlobal) {
                locale = language.getLocale(true)
            }
            return locale
        }

    override val melody: String
        get() = if (Module.isPro && !isGlobal) {
            prefs.birthdayMelody
        } else {
            prefs.melodyFile
        }

    override val isBirthdayInfiniteVibration: Boolean
        get() {
            var vibrate = prefs.isInfiniteVibrateEnabled
            if (Module.isPro && !isGlobal) {
                vibrate = prefs.isBirthdayInfiniteVibrationEnabled
            }
            return vibrate
        }

    override val isBirthdayInfiniteSound: Boolean
        get() {
            var isLooping = prefs.isInfiniteSoundEnabled
            if (Module.isPro && !isGlobal) {
                isLooping = prefs.isBirthdayInfiniteSoundEnabled
            }
            return isLooping
        }

    override val isVibrate: Boolean
        get() {
            var vibrate = prefs.isVibrateEnabled
            if (Module.isPro && !isGlobal) {
                vibrate = prefs.isBirthdayVibrationEnabled
            }
            return vibrate
        }

    override val uuId: String
        get() = if (mBirthday != null) {
            mBirthday!!.uuId
        } else
            ""

    override val id: Int
        get() = if (mBirthday != null) {
            mBirthday!!.uniqueId
        } else
            0

    override val ledColor: Int
        get() {
            var ledColor = LED.getLED(prefs.ledColor)
            if (Module.isPro && !isGlobal) {
                ledColor = LED.getLED(prefs.birthdayLedColor)
            }
            return ledColor
        }

    override val isAwakeDevice: Boolean
        get() {
            var isWake = prefs.isDeviceAwakeEnabled
            if (Module.isPro && !isGlobal) {
                isWake = prefs.isBirthdayWakeEnabled
            }
            return isWake
        }

    override val maxVolume: Int
        get() = prefs.loudness

    override val isGlobal: Boolean
        get() = prefs.isBirthdayGlobalEnabled

    override val isUnlockDevice: Boolean
        get() = prefs.isDeviceUnlockEnabled

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isScreenResumed = intent.getBooleanExtra(Constants.INTENT_NOTIFICATION, false)
        val key = intent.getIntExtra(Constants.INTENT_ID, 0)
        setContentView(R.layout.activity_show_birthday)
        card.setCardBackgroundColor(themeUtil.cardStyle)
        if (Module.isLollipop) {
            card.cardElevation = Configs.CARD_ELEVATION
        }
        loadImage(bgImage)
        colorify(buttonOk, buttonCall, buttonSend)

        buttonOk.onClick { ok() }
        buttonCall.onClick { call() }
        buttonSend.onClick { sendSMS() }

        buttonOk.setImageResource(R.drawable.ic_done_black_24dp)
        buttonCall.setImageResource(R.drawable.ic_call_black_24dp)
        buttonSend.setImageResource(R.drawable.ic_send_black_24dp)

        contactPhoto.borderColor = themeUtil.getColor(themeUtil.colorPrimary())
        contactPhoto.visibility = View.GONE

        initViewModel(key)
    }

    private fun initViewModel(id: Int) {
        viewModel = ViewModelProviders.of(this, BirthdayViewModel.Factory(application, id)).get(BirthdayViewModel::class.java)
        viewModel!!.birthday.observe(this, Observer<Birthday>{ birthday ->
            if (birthday != null) {
                showBirthday(birthday)
            }
        })
        viewModel!!.result.observe(this, Observer<Commands>{ commands ->
            if (commands != null) {
                when (commands) {
                    Commands.SAVED -> close()
                }
            }
        })
    }

    private fun showBirthday(birthday: Birthday) {
        this.mBirthday = birthday

        if (!TextUtils.isEmpty(birthday.number) && checkContactPermission()) {
            birthday.number = Contacts.getNumber(birthday.name, this)
        }
        if (birthday.contactId == 0 && !TextUtils.isEmpty(birthday.number) && checkContactPermission()) {
            birthday.contactId = Contacts.getIdFromNumber(birthday.number, this)
        }
        val photo = Contacts.getPhoto(birthday.contactId.toLong())
        if (photo != null) {
            contactPhoto.setImageURI(photo)
        } else {
            contactPhoto.visibility = View.GONE
        }
        val years = TimeUtil.getAgeFormatted(this, birthday.date)
        userName.text = birthday.name
        userName.contentDescription = birthday.name
        userYears.text = years
        userYears.contentDescription = years
        summary = birthday.name + "\n" + years
        if (TextUtils.isEmpty(birthday.number)) {
            buttonCall.hide()
            buttonSend.hide()
            userNumber.visibility = View.GONE
        } else {
            userNumber.text = birthday.number
            userNumber.contentDescription = birthday.number
        }
        showNotification(TimeUtil.getAge(birthday.date), birthday.name)
        if (isTtsEnabled) {
            startTts()
        }
    }

    private fun checkContactPermission(): Boolean {
        return Permissions.checkPermission(this, Permissions.READ_CONTACTS, Permissions.READ_CALLS)
    }

    private fun showNotification(years: Int, name: String) {
        val builder = NotificationCompat.Builder(this, Notifier.CHANNEL_REMINDER)
        builder.setContentTitle(name)
        builder.setContentText(TimeUtil.getAgeFormatted(this, years))
        if (Module.isLollipop) {
            builder.setSmallIcon(R.drawable.ic_cake_white_24dp)
            builder.color = ViewUtils.getColor(this, R.color.bluePrimary)
        } else {
            builder.setSmallIcon(R.drawable.ic_cake_nv_white)
        }
        if (!isScreenResumed && (!SuperUtil.isDoNotDisturbEnabled(this) || SuperUtil.checkNotificationPermission(this) && isBirthdaySilentEnabled)) {
            val sound = sound
            sound?.playAlarm(soundUri, isBirthdayInfiniteSound)
        }
        if (isVibrate) {
            var pattern = longArrayOf(150, 400, 100, 450, 200, 500, 300, 500)
            if (isBirthdayInfiniteVibration) {
                pattern = longArrayOf(150, 86400000)
            }
            builder.setVibrate(pattern)
        }
        if (Module.isPro) {
            builder.setLights(ledColor, 500, 1000)
        }
        val isWear = prefs.isWearEnabled
        if (isWear && Module.isJellyMR2) {
            builder.setOnlyAlertOnce(true)
            builder.setGroup("GROUP")
            builder.setGroupSummary(true)
        }
        val mNotifyMgr = NotificationManagerCompat.from(this)
        mNotifyMgr.notify(id, builder.build())
        if (isWear) {
            showWearNotification(name)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        removeFlags()
        if (prefs.isAutoBackupEnabled) {
            BackupTask(this).execute()
        }
        BackupBirthdaysTask(this).execute()
    }

    override fun onBackPressed() {
        discardMedia()
        if (prefs.isFoldingEnabled) {
            removeFlags()
            finish()
        } else {
            Toast.makeText(this@ShowBirthdayActivity, getString(R.string.select_one_of_item), Toast.LENGTH_SHORT).show()
        }
    }

    override fun call() {
        makeCall()
    }

    private fun makeCall() {
        if (Permissions.checkPermission(this, Permissions.CALL_PHONE) && mBirthday != null) {
            TelephonyUtil.makeCall(mBirthday!!.number, this)
            updateBirthday(mBirthday)
        } else {
            Permissions.requestPermission(this, CALL_PERM, Permissions.CALL_PHONE)
        }
    }

    override fun delay() {
        close()
    }

    override fun cancel() {
        sendSMS()
    }

    private fun sendSMS() {
        if (Permissions.checkPermission(this@ShowBirthdayActivity, Permissions.SEND_SMS) && mBirthday != null) {
            TelephonyUtil.sendSms(mBirthday!!.number, this@ShowBirthdayActivity)
            updateBirthday(mBirthday)
        } else {
            Permissions.requestPermission(this@ShowBirthdayActivity, SMS_PERM, Permissions.SEND_SMS)
        }
    }

    override fun favourite() {
        close()
    }

    override fun ok() {
        updateBirthday(mBirthday)
    }

    private fun updateBirthday(birthday: Birthday?) {
        if (birthday != null) {
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = System.currentTimeMillis()
            val year = calendar.get(Calendar.YEAR)
            birthday.showedYear = year
            viewModel!!.saveBirthday(birthday)
        }
    }

    private fun close() {
        removeFlags()
        discardNotification(id)
        finish()
    }

    override fun showSendingError() {
        buttonCall.setImageResource(R.drawable.ic_refresh)
        buttonCall.contentDescription = getString(R.string.acc_button_retry_to_send_message)
        if (buttonCall.visibility == View.GONE) {
            buttonCall.show()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.isEmpty()) return
        when (requestCode) {
            CALL_PERM -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                makeCall()
            }
            SMS_PERM -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                sendSMS()
            }
        }
    }

    companion object {

        private const val CALL_PERM = 612
        private const val SMS_PERM = 613

        fun getLaunchIntent(context: Context, id: Int): Intent {
            val resultIntent = Intent(context, ShowBirthdayActivity::class.java)
            resultIntent.putExtra(Constants.INTENT_ID, id)
            resultIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_MULTIPLE_TASK
            return resultIntent
        }
    }
}
